### PR TITLE
Middleware Runner

### PR DESCRIPTION
**Pull request by WyriHaximus**
Middleware Runner extracted from #213

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -136,6 +137,78 @@ $server->on('error', function (Exception $e) {
 Note that the request object can also emit an error.
 Check out [request](#request) for more details.
 
+### Middleware
+
+Middleware can ba added to the server using [`MiddlewareRunner`](src/MiddlewareRunner.php)
+instead of the the `callable`. A middleware is expected adjure to the following rules:
```
Comment (jsor) [inline README.md]: * the ~~the~~
* adjure -> to adhere (?)

File: README.md [hunk 2]
```diff
@@ -136,6 +137,78 @@ $server->on('error', function (Exception $e) {
...
+            return $next($request);
+        }
+        function (ServerRequestInterface $request, callable $next) {
+            return new Response(200);
+        }
+    ]
+));
+```
+
+#### Buffer
```
Comment (jsor) [inline README.md]: The `Buffer` and `LimitHandlers` are not part of this PR and the sections should probably removed.

File: README.md [hunk 3]
```diff
@@ -136,6 +137,78 @@ $server->on('error', function (Exception $e) {
...
 Check out [request](#request) for more details.
 
+### Middleware
+
+Middleware can ba added to the server using [`MiddlewareRunner`](src/MiddlewareRunner.php)
+instead of the the `callable`. A middleware is expected adjure to the following rules:
+
+* It is a `callable`.
+* It accepts `ServerRequestInterface` as first argument and `callable` as second argument.
+* It either returns a promise resolving to a `ResponseInterface`; or it calls the second argument
```
Comment (jsor) [inline README.md]: It is actually allowed to return a `ResponseInterface` directly (not only a Promise resolving to a `PromiseInterface`).

File: src/MiddlewareRunner.php [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        $cancel = null;
+        return new Promise\Promise(function ($resolve, $reject) use ($middleware, $request, $middlewareCollection, &$cancel, $defaultResponse) {
+            $cancel = $middleware(
+                $request,
+                new MiddlewareRunner(
+                    $defaultResponse,
+                    $middlewareCollection
+                )
+            );
+            Promise\resolve($cancel)->then($resolve, $reject);
```
Comment (jsor) [inline src/MiddlewareRunner.php]: This can be shortened to `$resolve($cancel)`.

File: README.md (hunk 2 above)
Comment (WyriHaximus) [reply README.md]: Yes correct, a PR for `Buffer` is coming soon and `LimitHandlers` after that. Then two more PR's will follow, one for the readme section and one for examples

File: README.md (hunk 3 above)
Comment (clue) [reply README.md]: How about this?

> * It is a `callable`.
> * It accepts `ServerRequestInterface` as first argument and optional `callable` as second argument.
> * It returns a `ResponseInterface` (or a `PromiseInterface` resolving to a `ResponseInterface`)
> * It calls `$next($request)` to continue processing the next middleware function or returns explicitly to abort the chain

File: src/MiddlewareRunner.php [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+     */
+    private $defaultResponse;
+
+    /**
+     * @var callable[]
+     */
+    private $middleware = array();
+
+    /**
+     * @param ResponseInterface $response
```
Comment (clue) [inline src/MiddlewareRunner.php]: See above :+1:

File: src/MiddlewareRunner.php [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        $defaultResponse = $this->defaultResponse;
+        $cancel = null;
+        return new Promise\Promise(function ($resolve, $reject) use ($middleware, $request, $middlewareCollection, &$cancel, $defaultResponse) {
+            $cancel = $middleware(
+                $request,
+                new MiddlewareRunner(
+                    $defaultResponse,
+                    $middlewareCollection
+                )
+            );
```
Comment (clue) [inline src/MiddlewareRunner.php]: I'm okay with this recursive implementation for now, but this is certainly something we would want to look into again in the future :+1:

File: README.md (hunk 3 above)
Comment (jsor) [reply README.md]: > * It returns a `ResponseInterface` (or a `PromiseInterface` resolving to a `ResponseInterface`)

I'd still write just *promise* because you can return any [thenable which can be consumed](https://github.com/reactphp/promise/blob/master/src/functions.php#L11) by react/promise.

Comment (clue) [reply README.md]: Valid point, just link to the promise repo and maybe provide a small example below? :+1:

Comment (jsor) [reply README.md]: Suggestion: 
> It returns a `ResponseInterface` (or any promise which can be consumed by [Promise\resolve()](https://github.com/reactphp/promise#resolve) resolving to a `ResponseInterface`)

Comment (clue) [reply README.md]: No objection, just wanted to point out that if probably makes sense to keep this in line with the first paragraph here: https://github.com/reactphp/http#response

Considering the other chapters, doesn't it make more sense to have this chapter *below* the response chapter? Middleware behave transparently to the `Server`, so most of what is described in "Response" still applies to middleware and the `Server` alike.

Comment (jsor) [reply README.md]: Makes sense 👍

File: README.md [hunk 7]
```diff
@@ -136,6 +137,78 @@ $server->on('error', function (Exception $e) {
...
+with a `ServerRequestInterface`.
+
+The following example adds a middleware that adds the current time to the request as a 
+header (`Request-Time`) and middleware that always returns a 200 code without a body: 
+
+```php
+$server = new Server(new MiddlewareRunner(
+    new Response(404), // Default response if no middleware returns a promise
```
[the remark this answers is not in the document]
Comment (WyriHaximus) [reply README.md]: What would you suggest the runner does when it doesn't have any middleware to run? Throw an exception?

File: src/MiddlewareRunner.php (hunk 6 above)
Comment (WyriHaximus) [reply src/MiddlewareRunner.php]: Yup, it works for now and I'll revisit it in the future :+1:

File: README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (WyriHaximus) [reply README.md]: Actually that would make sense as it's turned into a 500 because it is an internal server error \o/

**Comment by WyriHaximus** [comment]
Rebased and squashed the commits

File: README.md [hunk 8]
```diff
@@ -651,6 +652,31 @@ As such, HTTP/1.1 response messages will automatically include a
...
+
+The following example adds a middleware that adds the current time to the request as a 
+header (`Request-Time`) and middleware that always returns a 200 code without a body: 
+
+```php
+$server = new Server(new MiddlewareRunner([
+    function (ServerRequestInterface $request, callable $next) {
+        $request = $request->withHeader('Request-Time', time());
+        return $next($request);
+    }
```
[the remark this answers is not in the document]
Comment (WyriHaximus) [reply README.md]: Fixed :sunglasses:

File: README.md [hunk 9]
```diff
@@ -651,6 +652,31 @@ As such, HTTP/1.1 response messages will automatically include a
 `Connection: close` header, irrespective of what header values are
 passed explicitly.
 
+### Middleware
+
+Middleware can be added to the server using [`MiddlewareRunner`](src/MiddlewareRunner.php)
+instead of the `callable`. A middleware is expected to adjure the following rules:
```
Comment (clue) [inline README.md]: @jsor suggested `adhere` instead?

Comment (WyriHaximus) [reply README.md]: @clue whoops thought I fixed that one, fixed it, squashed it, and pushed it again :tada:
